hazelcast/src/main/java/com/hazelcast/transaction/impl/TransactionLog.java

```diff
@@ -38,53 +37,40 @@
  * the TransactionLog will contains 4 {@link TransactionLogRecord} instances.
  *
  * planned optimization:
- * Most transaction will be small, but an linkedlist + hashmap is created. Instead use an array and do a
+ * Most transaction will be small, but an linkedhashmap is created. Instead use an array and do a
  * linear search in that array. When there are too many items added, then enable the hashmap.
  */
 public class TransactionLog {
 
-    private final List<TransactionLogRecord> recordList = new LinkedList<TransactionLogRecord>();
-    private final Map<Object, TransactionLogRecord> recordMap = new HashMap<Object, TransactionLogRecord>();
+    private final Map<Object, TransactionLogRecord> recordMap = new LinkedHashMap<>();
 
     public TransactionLog() {
     }
 
     public TransactionLog(List<TransactionLogRecord> transactionLog) {
-        //
-        recordList.addAll(transactionLog);
+        for (TransactionLogRecord record : transactionLog) {
+            add(record);
+        }
     }
 
     public void add(TransactionLogRecord record) {
         Object key = record.getKey();
-
-        // there should be just one tx log for the same key. so if there is older we are removing it
-        if (key != null) {
-            TransactionLogRecord removed = recordMap.remove(key);
-            if (removed != null) {
-                recordList.remove(removed);
-            }
-        }
-
-        recordList.add(record);
-
-        if (key != null) {
-            recordMap.put(key, record);
+        if (key == null) {
+            key = new Object();
         }
+        recordMap.put(key, record);
     }
 
     public TransactionLogRecord get(Object key) {
         return recordMap.get(key);
     }
 
     public List<TransactionLogRecord> getRecordList() {
-        return recordList;
+        return new ArrayList<>(recordMap.values());
```

Comment:
Minor: do we need this copy here?

Reply:
This had kind of ripple effect, as the return type had to be changed from `List` to `Collection`, but I did that (see 16bee943c0e3ae0fbcbd690356959bd948a76880) and looks that everything's still good.